File: app/lib/services/omni_search_service.dart
import 'dart:math' as math;
import 'package:flutter/material.dart';

import '../config/theme.dart';
import '../models/metro_station.dart';
import '../models/pandal.dart';
import '../repositories/metro_repository.dart';
import '../repositories/supplementary_repository.dart';
import '../utils/constants.dart';
import '../utils/haversine.dart';
import 'pandal_search_service.dart';

/// Categories of search entities in the Universal Omni-Search
enum OmniCategory {
  all(label: 'All', icon: Icons.auto_awesome_rounded),
  pandals(label: 'Pandals', icon: Icons.temple_hindu_rounded),
  metro(label: 'Metro', icon: Icons.subway_rounded),
  food(label: 'Food', icon: Icons.restaurant_rounded);

  const OmniCategory({required this.label, required this.icon});
  final String label;
  final IconData icon;
}

/// Type of matched search entity
enum OmniResultType {
  pandal,
  metro,
  food,
}

/// Unified auto-complete result entity returned by [OmniSearchService]
class OmniSearchResult {
  const OmniSearchResult({
    required this.type,
    required this.id,
    required this.title,
    required this.subtitle,
    required this.latitude,
    required this.longitude,
    required this.score,
    required this.highlightSpans,
    this.matchedField = 'Name',
    this.distanceMeters,
    this.pandal,
    this.metroStation,
    this.foodSpot,
    this.badgeText,
    this.badgeColor,
    this.rating,
  });

  final OmniResultType type;
  final String id;
  final String title;
  final String subtitle;
  final double latitude;
  final double longitude;
  final double score;
  final List<TextSpan> highlightSpans;
  final String matchedField;
  final double? distanceMeters;
  final Pandal? pandal;
  final MetroStation? metroStation;
  final FoodSpot? foodSpot;
  final String? badgeText;
  final Color? badgeColor;
  final double? rating;

  IconData get icon {
    switch (type) {
      case OmniResultType.pandal:
        return Icons.temple_hindu_rounded;
      case OmniResultType.metro:
        return Icons.subway_rounded;
      case OmniResultType.food:
        return Icons.restaurant_rounded;
    }
  }

  Color get accentColor {
    switch (type) {
      case OmniResultType.pandal:
        return PujaColors.durgaRed;
      case OmniResultType.metro:
        return metroStation?.line.color ?? const Color(0xFF1976D2);
      case OmniResultType.food:
        return const Color(0xFFFF9100); // warm culinary amber
    }
  }
}

/// Unified system-wide search engine (inspired by Rofi / Spotlight / Arch Linux Omachi)
/// Searches simultaneously across Pandals, Metro Stations, and Culinary Spots.
class OmniSearchService {
  OmniSearchService._();
  static final OmniSearchService instance = OmniSearchService._();

  // In-memory cache of recent search keywords
  final List<String> _recentSearches = [
    'Ekdalia Evergreen',
    'Shyambazar Metro',
    'Golbari',
    'Mitra Cafe',
    'College Square',
    'Esplanade',
  ];

  List<String> get recentSearches => List.unmodifiable(_recentSearches);

  void addRecentSearch(String query) {
    final trimmed = query.trim();
    if (trimmed.isEmpty) return;
    _recentSearches.removeWhere((item) => item.toLowerCase() == trimmed.toLowerCase());
    _recentSearches.insert(0, trimmed);
    if (_recentSearches.length > 8) {
      _recentSearches.removeLast();
    }
  }

  void clearRecentSearches() {
    _recentSearches.clear();
  }

  /// Evaluates and scores a Metro Station against a search query
  OmniSearchResult? scoreMetroStation(
    MetroStation station,
    String rawQuery, {
    double? userLat,
    double? userLng,
    TextStyle? normalStyle,
    TextStyle? highlightStyle,
  }) {
    final query = rawQuery.trim().toLowerCase();
    if (query.isEmpty) return null;

    final normQuery = PandalSearchService.normalize(query);
    final queryTokens = normQuery.split(' ').where((t) => t.isNotEmpty).toList();
    if (queryTokens.isEmpty) return null;

    final normName = PandalSearchService.normalize(station.name);
    final normLine = PandalSearchService.normalize(station.line.label);
    final normCorridor = PandalSearchService.normalize(station.line.corridor);
    final normAliases = station.aliases.map(PandalSearchService.normalize).toList();
    final normNearby = station.popularPandalsNearby
        .map(PandalSearchService.normalize)
        .join(' ');

    double score = 0.0;
    String matchedField = 'Name';

    // 1. Exact or prefix match on station name or aliases
    if (normName == normQuery || normAliases.contains(normQuery)) {
      score += 1300;
    } else if (normName.startsWith(normQuery) || normAliases.any((a) => a.startsWith(normQuery))) {
      score += 850;
    } else if (normName.contains(normQuery) || normAliases.any((a) => a.contains(normQuery))) {
      score += 480;
    }

    // 2. Token evaluation across line & nearby pandals
    int tokensMatched = 0;
    for (final token in queryTokens) {
      bool tokenMatched = false;

      // Word match in name
      final nameWords = normName.split(' ');
      for (final w in nameWords) {
        if (w == token) {
          score += 320;
          tokenMatched = true;
          break;
        } else if (w.startsWith(token)) {
          score += 220;
          tokenMatched = true;
          break;
        }
      }

      if (!tokenMatched && normName.contains(token)) {
        score += 160;
        tokenMatched = true;
      }

      // Check "metro" keyword
      if (token == 'metro' || token == 'station') {
        score += 200;
        tokenMatched = true;
        matchedField = 'Metro System';
      }

      // Check line or corridor
      if (normLine.contains(token) || normCorridor.contains(token)) {
        score += 180;
        tokenMatched = true;
        matchedField = 'Metro Line';
      }

      // Check nearby famous pandals connected to this station
      if (normNearby.contains(token)) {
        score += 150;
        tokenMatched = true;
        matchedField = 'Nearby Pandal';
      }

      if (tokenMatched) tokensMatched++;
    }

    if (tokensMatched < queryTokens.length) {
      return null;
    }

    // Interchange bonus
    if (station.isInterchange) {
      score += 50;
    }

    // Distance bonus
    double? distMeters;
    if (userLat != null && userLng != null) {
      distMeters = haversineMeters(userLat, userLng, station.latitude, station.longitude);
      final distKm = distMeters / 1000.0;
      score += math.max(0.0, (15.0 - distKm) * 5.0);
    }

    final spans = PandalSearchService.buildHighlightSpans(
      text: station.name,
      query: query,
      normalStyle: normalStyle,
      highlightStyle: highlightStyle,
    );

    return OmniSearchResult(
      type: OmniResultType.metro,
      id: station.id,
      title: station.name,
      subtitle: station.subtitle,
      latitude: station.latitude,
      longitude: station.longitude,
      score: score,
      highlightSpans: spans,
      matchedField: matchedField,
      distanceMeters: distMeters,
      metroStation: station,
      badgeText: station.line.code.toUpperCase(),
      badgeColor: station.line.color,
    );
  }

  /// Evaluates and scores a Food Spot / Restaurant against a search query
  OmniSearchResult? scoreFoodSpot(
    FoodSpot spot,
    String rawQuery, {
    double? userLat,
    double? userLng,
    TextStyle? normalStyle,
    TextStyle? highlightStyle,
  }) {
    final query = rawQuery.trim().toLowerCase();
    if (query.isEmpty) return null;

    final normQuery = PandalSearchService.normalize(query);
    final queryTokens = normQuery.split(' ').where((t) => t.isNotEmpty).toList();
    if (queryTokens.isEmpty) return null;

    final normName = PandalSearchService.normalize(spot.name);
    final normType = PandalSearchService.normalize(spot.type);
    final normMustTry = PandalSearchService.normalize(spot.mustTry ?? '');
    final normNearby = PandalSearchService.normalize(spot.nearbyPandal);

    double score = 0.0;
    String matchedField = 'Name';

    // 1. Exact or prefix match on name
    if (normName == normQuery) {
      score += 1250;
    } else if (normName.startsWith(normQuery)) {
      score += 820;
    } else if (normName.contains(normQuery)) {
      score += 460;
    }

    // 2. Token evaluation across cuisine, must-try items, and location
    int tokensMatched = 0;
    for (final token in queryTokens) {
      bool tokenMatched = false;

      final words = normName.split(' ');
      for (final w in words) {
        if (w == token) {
          score += 310;
          tokenMatched = true;
          break;
        } else if (w.startsWith(token)) {
          score += 210;
          tokenMatched = true;
          break;
        }
      }

      if (!tokenMatched && normName.contains(token)) {
        score += 150;
        tokenMatched = true;
      }

      // Check cuisine / type (e.g. Biryani, Rolls, Snacks, Sweets)
      if (normType.contains(token)) {
        score += 180;
        tokenMatched = true;
        matchedField = 'Cuisine';
      }

      // Check must-try specialty (e.g. Kosha Mangsho, Fish Fry, Kabiraji)
      if (normMustTry.contains(token)) {
        score += 190;
        tokenMatched = true;
        matchedField = 'Specialty';
      }

      // Check nearby pandal
      if (normNearby.contains(token)) {
        score += 140;
        tokenMatched = true;
        matchedField = 'Near Pandal';
      }

      // Food keywords
      if (token == 'food' || token == 'restaurant' || token == 'eat' || token == 'eating') {
        score += 120;
        tokenMatched = true;
        matchedField = 'Food Spot';
      }

      if (tokenMatched) tokensMatched++;
    }

    if (tokensMatched < queryTokens.length) {
      return null;
    }

    // Distance bonus
    double? distMeters;
    if (userLat != null && userLng != null) {
      distMeters = haversineMeters(userLat, userLng, spot.lat, spot.lng);
      final distKm = distMeters / 1000.0;
      score += math.max(0.0, (15.0 - distKm) * 5.0);
    }

    // Rating boost
    if (spot.rating != null) {
      score += (spot.rating! * 5.0);
    }

    final spans = PandalSearchService.buildHighlightSpans(
      text: spot.name,
      query: query,
      normalStyle: normalStyle,
      highlightStyle: highlightStyle,
    );

    final subtitleParts = <String>[];
    if (spot.mustTry != null && spot.mustTry!.isNotEmpty) {
      subtitleParts.add(spot.mustTry!);
    } else {
      subtitleParts.add(spot.type);
    }
    if (spot.nearbyPandal.isNotEmpty) {
      subtitleParts.add('Near ${spot.nearbyPandal}');
    }

    return OmniSearchResult(
      type: OmniResultType.food,
      id: spot.id,
      title: spot.name,
      subtitle: subtitleParts.join(' · '),
      latitude: spot.lat,
      longitude: spot.lng,
      score: score,
      highlightSpans: spans,
      matchedField: matchedField,
      distanceMeters: distMeters,
      foodSpot: spot,
      badgeText: spot.priceRange ?? 'FOOD',
      badgeColor: const Color(0xFFFF9100),
      rating: spot.rating,
    );
  }

  /// Converts a PandalSearchResult to an OmniSearchResult
  OmniSearchResult _fromPandalResult(PandalSearchResult res) {
    final p = res.pandal;
    final subtitleParts = <String>[p.zone.label];
    if (p.nearestMetro != null && p.nearestMetro!.isNotEmpty) {
      subtitleParts.add('🚇 ${p.nearestMetro}');
    }
    if (p.theme.isNotEmpty) {
      subtitleParts.add(p.theme);
    }

    return OmniSearchResult(
      type: OmniResultType.pandal,
      id: p.id,
      title: p.name,
      subtitle: subtitleParts.join(' · '),
      latitude: p.lat,
      longitude: p.lng,
      score: res.score,
      highlightSpans: res.nameHighlightSpans,
      matchedField: res.matchedField,
      distanceMeters: res.distanceMeters,
      pandal: p,
      badgeText: p.zone.shortLabel,
      badgeColor: PujaColors.durgaRed,
      rating: p.rating,
    );
  }

  /// Performs a universal query across all system entities:
  /// Pandals, Metro Stations, and Food Spots.
  List<OmniSearchResult> search({
    required String query,
    required List<Pandal> pandals,
    List<FoodSpot>? foodSpots,
    List<MetroStation>? metroStations,
    OmniCategory category = OmniCategory.all,
    int limit = 8,
    double? userLat,
    double? userLng,
    TextStyle? normalStyle,
    TextStyle? highlightStyle,
  }) {
    final trimmed = query.trim();
    if (trimmed.isEmpty) return const [];

    final stations = metroStations ?? MetroRepository.allStations;
    final results = <OmniSearchResult>[];

    // 1. Search Pandals
    if (category == OmniCategory.all || category == OmniCategory.pandals) {
      for (final p in pandals) {
        final res = PandalSearchService.instance.scorePandal(
          p,
          trimmed,
          userLat: userLat,
          userLng: userLng,
          normalStyle: normalStyle,
          highlightStyle: highlightStyle,
        );
        if (res != null) {
          results.add(_fromPandalResult(res));
        }
      }
    }

    // 2. Search Metro Stations
    if (category == OmniCategory.all || category == OmniCategory.metro) {
      for (final station in stations) {
        final res = scoreMetroStation(
          station,
          trimmed,
          userLat: userLat,
          userLng: userLng,
          normalStyle: normalStyle,
          highlightStyle: highlightStyle,
        );
        if (res != null) {
          results.add(res);
        }
      }
    }

    // 3. Search Food Spots
    if ((category == OmniCategory.all || category == OmniCategory.food) && foodSpots != null) {
      for (final spot in foodSpots) {
        final res = scoreFoodSpot(
          spot,
          trimmed,
          userLat: userLat,
          userLng: userLng,
          normalStyle: normalStyle,
          highlightStyle: highlightStyle,
        );
        if (res != null) {
          results.add(res);
        }
      }
    }

    // Sort by relevance score descending
    results.sort((a, b) => b.score.compareTo(a.score));

    if (results.length > limit) {
      return results.sublist(0, limit);
    }
    return results;
  }
}

File: ios/Uma/Services/OmniSearchService.swift
// Services/OmniSearchService.swift — port of services/omni_search_service.dart
// Phonetic / token-weighted bilingual search across Pandals, Metro, Food.

import Foundation
import SwiftUI
import CoreLocation

enum OmniCategory: String, CaseIterable, Identifiable {
    case all, pandals, metro, food
    var id: String { rawValue }
    var label: String {
        switch self {
        case .all:     return "All"
        case .pandals: return "Pandals"
        case .metro:   return "Metro"
        case .food:    return "Food"
        }
    }
    var iconName: String {
        switch self {
        case .all:     return "sparkles"
        case .pandals: return "building.columns"
        case .metro:   return "tram.fill"
        case .food:    return "fork.knife"
        }
    }
}

enum OmniResultType: String, Codable, Sendable {
    case pandal, metro, food
}

struct OmniSearchResult: Identifiable, Hashable, Sendable {
    let type: OmniResultType
    let id: String
    let title: String
    let subtitle: String
    let latitude: Double
    let longitude: Double
    let score: Double
    let matchedField: String
    let distanceMeters: Double?
    let badgeText: String?
    let badgeColorHex: UInt32?
    let rating: Double?
    let pandalId: String?
    let metroStationId: String?
    let foodSpotId: String?

    var iconName: String {
        switch type {
        case .pandal: return "building.columns"
        case .metro:  return "tram.fill"
        case .food:   return "fork.knife"
        }
    }

    var badgeColor: Color? {
        guard let h = badgeColorHex else { return nil }
        return Color(hex: h)
    }

    var coordinate: CLLocationCoordinate2D {
        .init(latitude: latitude, longitude: longitude)
    }
}

/// Normalizes text for comparison: lowercase, strips punctuation, collapses whitespace.
enum SearchNormalize {
    static func normalize(_ raw: String) -> String {
        let lowered = raw.lowercased()
        let scalars = lowered.unicodeScalars.map { scalar -> Character in
            CharacterSet.alphanumerics.contains(scalar) ? Character(scalar) : " "
        }
        let collapsed = String(scalars).split(separator: " ", omittingEmptySubsequences: true)
            .joined(separator: " ")
        return collapsed
    }

    static func buildHighlightSpans(text: String, query: String,
                                    normalStyle: AttributedString = .init(),
                                    highlightStyle: AttributedString = .init()) -> AttributedString {
        var result = AttributedString(text)
        let q = query.trimmingCharacters(in: .whitespaces).lowercased()
        guard !q.isEmpty else { return result }
        let lower = text.lowercased()
        var searchStart = lower.startIndex
        while let range = lower.range(of: q, range: searchStart..<lower.endIndex) {
            if let attrRange = Range(range, in: result) {
                result[attrRange].foregroundColor = .primary
            }
            searchStart = range.upperBound
        }
        _ = (normalStyle, highlightStyle)
        return result
    }
}

final class OmniSearchService: @unchecked Sendable {
    static let shared = OmniSearchService()
    private init() {}

    private var searchCache: [String: [OmniSearchResult]] = [:]
    private let cacheLimit = 80

    /// Search across pandals, food spots, and metro stations.
    /// Default cap of 8 results per query, like the Flutter source.
    func search(
        query: String,
        pandals: [Pandal],
        foodSpots: [FoodSpot] = [],
        metroStations: [MetroStation] = [],
        category: OmniCategory = .all,
        limit: Int = 8,
        userLat: Double? = nil,
        userLng: Double? = nil
    ) -> [OmniSearchResult] {
        let trimmed = query.trimmingCharacters(in: .whitespaces)
        guard !trimmed.isEmpty else { return [] }

        let cacheKey = "\(category.rawValue)|\(trimmed.lowercased())|\(limit)|\(pandals.count)|\(foodSpots.count)|\(metroStations.count)|\(userLat?.description ?? "")|\(userLng?.description ?? "")"
        if let cached = searchCache[cacheKey] { return cached }

        var results: [OmniSearchResult] = []
        if category == .all || category == .pandals {
            for p in pandals {
                if let r = scorePandal(p, query: trimmed, userLat: userLat, userLng: userLng) {
                    results.append(r)
                }
            }
        }
        if category == .all || category == .metro {
            for st in metroStations {
                if let r = scoreMetroStation(st, query: trimmed, userLat: userLat, userLng: userLng) {
                    results.append(r)
                }
            }
        }
        if (category == .all || category == .food), !foodSpots.isEmpty {
            for f in foodSpots {
                if let r = scoreFoodSpot(f, query: trimmed, userLat: userLat, userLng: userLng) {
                    results.append(r)
                }
            }
        }
        results.sort { $0.score > $1.score }
        let final = Array(results.prefix(limit))

        if searchCache.count >= cacheLimit { searchCache.removeAll() }
        searchCache[cacheKey] = final
        return final
    }

    func clearCache() { searchCache.removeAll() }

    // MARK: - Scoring

    func scorePandal(_ p: Pandal, query: String, userLat: Double? = nil, userLng: Double? = nil) -> OmniSearchResult? {
        let q = SearchNormalize.normalize(query)
        let tokens = q.split(separator: " ").map(String.init).filter { !$0.isEmpty }
        guard !tokens.isEmpty else { return nil }

        let normName = SearchNormalize.normalize(p.name)
        let normTheme = SearchNormalize.normalize(p.theme)
        let normArea = SearchNormalize.normalize(p.area ?? "")
        let normMetro = SearchNormalize.normalize(p.nearestMetro ?? "")

        var score: Double = 0
        var matchedField = "Name"

        if normName == q { score += 1300 }
        else if normName.hasPrefix(q) { score += 850 }
        else if normName.contains(q) { score += 480 }

        var tokensMatched = 0
        for t in tokens {
            var matched = false
            for w in normName.split(separator: " ").map(String.init) {
                if w == t { score += 320; matched = true; break }
                if w.hasPrefix(t) { score += 220; matched = true; break }
            }
            if !matched && normName.contains(t) { score += 160; matched = true }
            if !matched && (normTheme.contains(t) || normArea.contains(t) || normMetro.contains(t)) {
                score += 140
                matched = true
                matchedField = "Theme / Metro"
            }
            if matched { tokensMatched += 1 }
        }
        guard tokensMatched >= tokens.count else { return nil }

        // Rating boost
        if let r = p.rating { score += r * 5 }

        // Distance bonus
        var distMeters: Double?
        if let ulat = userLat, let ulng = userLng {
            let d = haversineMeters(ulat, ulng, p.lat, p.lng)
            distMeters = d
            let km = d / 1000
            score += max(0, (15 - km) * 5)
        }

        var subtitleParts: [String] = [p.zone.label]
        if let m = p.nearestMetro, !m.isEmpty { subtitleParts.append("🚇 \(m)") }
        if !p.theme.isEmpty { subtitleParts.append(p.theme) }

        return OmniSearchResult(
            type: .pandal,
            id: p.id,
            title: p.name,
            subtitle: subtitleParts.joined(separator: " · "),
            latitude: p.lat,
            longitude: p.lng,
            score: score,
            matchedField: matchedField,
            distanceMeters: distMeters,
            badgeText: p.zone.shortLabel.uppercased(),
            badgeColorHex: 0xB71C1C,
            rating: p.rating,
            pandalId: p.id,
            metroStationId: nil,
            foodSpotId: nil)
    }

    func scoreMetroStation(_ s: MetroStation, query: String, userLat: Double? = nil, userLng: Double? = nil) -> OmniSearchResult? {
        let q = SearchNormalize.normalize(query)
        let tokens = q.split(separator: " ").map(String.init).filter { !$0.isEmpty }
        guard !tokens.isEmpty else { return nil }

        let normName = SearchNormalize.normalize(s.name)
        let normLine = SearchNormalize.normalize(s.line.label)
        let normCorridor = SearchNormalize.normalize(s.line.corridor)
        let normAliases = s.aliases.map { SearchNormalize.normalize($0) }

        var score: Double = 0
        var matchedField = "Name"

        if normName == q || normAliases.contains(q) { score += 1300 }
        else if normName.hasPrefix(q) || normAliases.contains(where: { $0.hasPrefix(q) }) { score += 850 }
        else if normName.contains(q) || normAliases.contains(where: { $0.contains(q) }) { score += 480 }

        var tokensMatched = 0
        for t in tokens {
            var matched = false
            for w in normName.split(separator: " ").map(String.init) {
                if w == t { score += 320; matched = true; break }
                if w.hasPrefix(t) { score += 220; matched = true; break }
            }
            if !matched && normName.contains(t) { score += 160; matched = true }
            if t == "metro" || t == "station" {
                score += 200; matched = true; matchedField = "Metro System"
            }
            if normLine.contains(t) || normCorridor.contains(t) {
                score += 180; matched = true; matchedField = "Metro Line"
            }
            if matched { tokensMatched += 1 }
        }
        guard tokensMatched >= tokens.count else { return nil }

        if s.isInterchange { score += 50 }

        var distMeters: Double?
        if let ulat = userLat, let ulng = userLng {
            let d = haversineMeters(ulat, ulng, s.latitude, s.longitude)
            distMeters = d
            let km = d / 1000
            score += max(0, (15 - km) * 5)
        }

        return OmniSearchResult(
            type: .metro,
            id: s.id,
            title: s.name,
            subtitle: s.subtitle,
            latitude: s.latitude,
            longitude: s.longitude,
            score: score,
            matchedField: matchedField,
            distanceMeters: distMeters,
            badgeText: s.line.code.uppercased(),
            badgeColorHex: hexFromColor(s.line.color),
            rating: nil,
            pandalId: nil,
            metroStationId: s.id,
            foodSpotId: nil)
    }

    func scoreFoodSpot(_ f: FoodSpot, query: String, userLat: Double? = nil, userLng: Double? = nil) -> OmniSearchResult? {
        let q = SearchNormalize.normalize(query)
        let tokens = q.split(separator: " ").map(String.init).filter { !$0.isEmpty }
        guard !tokens.isEmpty else { return nil }

        let normName = SearchNormalize.normalize(f.name)
        let normType = SearchNormalize.normalize(f.type)
        let normMustTry = SearchNormalize.normalize(f.mustTry ?? "")

        var score: Double = 0
        var matchedField = "Name"

        if normName == q { score += 1250 }
        else if normName.hasPrefix(q) { score += 820 }
        else if normName.contains(q) { score += 460 }

        var tokensMatched = 0
        for t in tokens {
            var matched = false
            for w in normName.split(separator: " ").map(String.init) {
                if w == t { score += 310; matched = true; break }
                if w.hasPrefix(t) { score += 210; matched = true; break }
            }
            if !matched && normName.contains(t) { score += 150; matched = true }
            if normType.contains(t) { score += 180; matched = true; matchedField = "Cuisine" }
            if normMustTry.contains(t) { score += 190; matched = true; matchedField = "Specialty" }
            if t == "food" || t == "restaurant" || t == "eat" || t == "eating" {
                score += 120; matched = true; matchedField = "Food Spot"
            }
            if matched { tokensMatched += 1 }
        }
        guard tokensMatched >= tokens.count else { return nil }

        if let r = f.rating { score += r * 5 }

        var distMeters: Double?
        if let ulat = userLat, let ulng = userLng {
            let d = haversineMeters(ulat, ulng, f.lat, f.lng)
            distMeters = d
            let km = d / 1000
            score += max(0, (15 - km) * 5)
        }

        var subtitleParts: [String] = []
        if let m = f.mustTry, !m.isEmpty { subtitleParts.append(m) } else { subtitleParts.append(f.type) }
        if !f.nearbyPandal.isEmpty { subtitleParts.append("Near \(f.nearbyPandal)") }

        return OmniSearchResult(
            type: .food,
            id: f.id,
            title: f.name,
            subtitle: subtitleParts.joined(separator: " · "),
            latitude: f.lat,
            longitude: f.lng,
            score: score,
            matchedField: matchedField,
            distanceMeters: distMeters,
            badgeText: f.priceRange ?? "FOOD",
            badgeColorHex: 0xFF9100,
            rating: f.rating,
            pandalId: nil,
            metroStationId: nil,
            foodSpotId: f.id)
    }
}

/// Extract a hex UInt32 from a SwiftUI Color (best-effort, for color badges).
private func hexFromColor(_ color: Color) -> UInt32 {
    let resolved = color.resolve(in: .init())
    let r = UInt32(max(0, min(255, Int(resolved.red * 255))))
    let g = UInt32(max(0, min(255, Int(resolved.green * 255))))
    let b = UInt32(max(0, min(255, Int(resolved.blue * 255))))
    return (r << 16) | (g << 8) | b
}
